lib/eclipse/EclipseState/Grid/GridProperty.cpp

```diff
@@ -462,19 +462,13 @@ std::vector< double > temperature_lookup( size_t size,
                                           const GridProperties<int>* ig_props ) {
 
     if (tables->hasTables("RTEMPVD")) {
-        if( !tables->useEqlnum() ) {
-            /* if values are defaulted in the TEMPI keyword, but no
-             * EQLNUM is specified, you will get NaNs.
-             */
-            return std::vector< double >( size, std::numeric_limits< double >::quiet_NaN() );
-        }
+        const std::vector< int >& eqlNum = ig_props->getKeyword("EQLNUM").getData();
 
         const auto& rtempvdTables = tables->getRtempvdTables();
-        const std::vector< int >& eqlNum = ig_props->getKeyword("EQLNUM").getData();
         std::vector< double > values( size, 0 );
 
         for (size_t cellIdx = 0; cellIdx < eqlNum.size(); ++ cellIdx) {
-            int cellEquilNum = eqlNum[cellIdx];
+            int cellEquilNum = eqlNum[cellIdx] - 1; // EQLNUM contains fortran-style indices!
```

Comment:
No; the `EQLNUM` values are used as ID's and not indices; so this line should be removed.

Reply:
are you sure? IIRC, this caused a segmentation fault for me. Also, this array just seems to contain the data provided by the EQLNUM grid propery which uses fortran-style indices. I would not have bothered if it didn't crash on me.

Reply:
maybe the variable should be renamed to `cellEquilRegionIdx` because an index is needed when retrieving the corresponding table below. done.

Reply:
>  are you sure? IIRC, this caused a segmentation fault for me.

yes - I am sure that the `EQLNUM` values are used as ID's and not indices. If things crash for you that is obviously a bug somewhere, and we should fix that - but this is not the right fix.

We don't really understand eachother: 

Joakim:
> the EQLNUM values are used as ID's and not indices

Andreas:
> maybe the variable should be renamed to cellEquilRegionIdx because an index is needed

Reply:
> yes - I am sure that the EQLNUM values are used as ID's and not indices.

true.

> the EQLNUM values are used as ID's and not indices

a C-style index needs to be passed to `rtempvdTables.getTable()` if I'm not horribly mistaken? I agree that in the version of the PR which you reviewed, the variable name was misleading.

Reply:
> a C-style index needs to be passed to rtempvdTables.getTable() if I'm not horribly mistaken? 

Well - you are horribly [mistaken](https://github.com/OPM/opm-parser/blob/master/lib/eclipse/EclipseState/Tables/TableContainer.cpp#L51)

Reply:
am I?

if I revert these hunks I get an exception:

```
TableContainer - invalid tableNumber. Abort!
```

while it works fine if I leave them in place. maybe the code you pointed to is not the one which is actually called?

Reply:
As I said I am not ruling out a bug somewhere, but I do not believe in your fix. Please provide a backtrace.

Reply:
you are not the most trusting person, are you? ;)

here's the backtrace:

```
(gdb) catch throw
Catchpoint 1 (throw)
(gdb) r
Starting program: /home/and/src/ewoms/build-cmake/bin/ebos --ecl-deck-file-name=/home/and/tmp/spe1-thermal/SPE1CASE2_THERMAL_ROCKHEATCAP.DATA --newton-max-error=1000 --enable-ecl-output=false --enable-vtk-output=true
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffef908700 (LWP 11411)]
eWoms 2018.04-pre ("Winston") will now start the trip. Please sit back, relax and enjoy the ride.
Allocating the grid
Reading the deck file '/home/and/tmp/spe1-thermal/SPE1CASE2_THERMAL_ROCKHEATCAP.DATA'
Distributing the grid
Allocating the model
Allocating the problem
Finish init of the model
Finish init of the problem

Thread 1 "ebos" hit Catchpoint 1 (exception thrown), 0x00007ffff3e768bd in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff3e768bd in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00000000007aed91 in Opm::TableContainer::getTable(unsigned long) const ()
#2  0x000000000075fa06 in Opm::temperature_lookup(unsigned long, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int> const*) ()
#3  0x0000000000748383 in std::_Function_handler<std::vector<double, std::allocator<double> > (unsigned long), std::_Bind<std::vector<double, std::allocator<double> > (*(std::_Placeholder<1>, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int>*))(unsigned long, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int> const*)> >::_M_invoke(std::_Any_data const&, unsigned long&&) ()
#4  0x0000000000762abf in Opm::GridProperty<double>::GridProperty(unsigned long, unsigned long, unsigned long, Opm::GridPropertySupportedKeywordInfo<double> const&) ()
#5  0x00000000007590f8 in Opm::GridProperties<double>::insertKeyword(Opm::GridPropertySupportedKeywordInfo<double> const&) const ()
#6  0x0000000000759816 in Opm::GridProperties<double>::addAutoGeneratedKeyword_(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const ()
#7  0x0000000000759bb3 in Opm::GridProperties<double>::getKeyword(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#8  0x0000000000742a6d in Opm::Eclipse3DProperties::getDoubleGridProperty(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const ()
#9  0x00000000006a64ad in Ewoms::EclEquilInitializer<Ewoms::Properties::TTag::EclProblem>::EclEquilInitializer<Opm::EclMaterialLawManager<Opm::ThreePhaseMaterialTraits<double, 0, 1, 2> > >(Ewoms::Simulator<Ewoms::Properties::TTag::EclProblem> const&, Opm::EclMaterialLawManager<Opm::ThreePhaseMaterialTraits<double, 0, 1, 2> >&) ()
#10 0x00000000006e6edc in Ewoms::EclProblem<Ewoms::Properties::TTag::EclProblem>::finishInit() ()
#11 0x00000000006e8ec5 in Ewoms::Simulator<Ewoms::Properties::TTag::EclProblem>::Simulator(bool) ()
#12 0x000000000059549b in int Ewoms::start<Ewoms::Properties::TTag::EclProblem>(int, char**) ()
#13 0x00007ffff3303830 in __libc_start_main (main=0x579930 <main>, argc=5, argv=0x7fffffffe778, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffe768) at ../csu/libc-start.c:291
#14 0x0000000000591399 in _start ()
```

Reply:
and once again with debugging symbols and no optimization:

```
(gdb) catch throw
Catchpoint 1 (throw)
(gdb) r
Starting program: /home/and/src/ewoms/build-cmake/bin/ebos --ecl-deck-file-name=/home/and/tmp/spe1-thermal/SPE1CASE2_THERMAL_ROCKHEATCAP.DATA --newton-max-error=1000 --enable-ecl-output=false --enable-vtk-output=true
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7fffef908700 (LWP 6867)]
eWoms 2018.04-pre ("Winston") will now start the trip. Please sit back, relax and enjoy the ride.
Allocating the grid
Reading the deck file '/home/and/tmp/spe1-thermal/SPE1CASE2_THERMAL_ROCKHEATCAP.DATA'
Distributing the grid
Allocating the model
Allocating the problem
Finish init of the model
Finish init of the problem

Thread 1 "ebos" hit Catchpoint 1 (exception thrown), 0x00007ffff3e768bd in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff3e768bd in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000000000f2a0be in Opm::TableContainer::getTable (this=0x198c820, tableNumber=1)
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Tables/TableContainer.cpp:53
#2  0x0000000000eab9c3 in Opm::TableContainer::getTable<Opm::RtempvdTable> (this=0x198c820, tableNumber=1)
    at /home/and/src/opm-parser/lib/eclipse/include/opm/parser/eclipse/EclipseState/Tables/TableContainer.hpp:82
#3  0x0000000000ea6afe in Opm::temperature_lookup (size=300, tables=0x1988dd0, grid=0x19891e0, ig_props=0x1989310)
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Grid/GridProperty.cpp:472
#4  0x0000000000e85917 in std::_Bind<std::vector<double, std::allocator<double> > (*(std::_Placeholder<1>, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int>*))(unsigned long, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int> const*)>::__call<std::vector<double, std::allocator<double> >, unsigned long&&, 0ul, 1ul, 2ul, 3ul>(std::tuple<unsigned long&&>&&, std::_Index_tuple<0ul, 1ul, 2ul, 3ul>) (this=0x1957680, 
    __args=<unknown type in /home/and/src/ewoms/build-cmake/bin/ebos, CU 0xaaaf07, DIE 0xb1a7ae>) at /usr/include/c++/5/functional:1074
#5  0x0000000000e85132 in std::_Bind<std::vector<double, std::allocator<double> > (*(std::_Placeholder<1>, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int>*))(unsigned long, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int> const*)>::operator()<unsigned long, std::vector<double, std::allocator<double> > >(unsigned long&&) (this=0x1957680) at /usr/include/c++/5/functional:1133
#6  0x0000000000e84854 in std::_Function_handler<std::vector<double, std::allocator<double> > (unsigned long), std::_Bind<std::vector<double, std::allocator<double> > (*(std::_Placeholder<1>, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int>*))(unsigned long, Opm::TableManager const*, Opm::EclipseGrid const*, Opm::GridProperties<int> const*)> >::_M_invoke(std::_Any_data const&, unsigned long&&) (__functor=..., 
    __args#0=<unknown type in /home/and/src/ewoms/build-cmake/bin/ebos, CU 0xaaaf07, DIE 0xb1706b>) at /usr/include/c++/5/functional:1857
#7  0x0000000000eabd30 in std::function<std::vector<double, std::allocator<double> > (unsigned long)>::operator()(unsigned long) const (this=0x194e718, 
    __args#0=300) at /usr/include/c++/5/functional:2267
#8  0x0000000000ea9f5c in Opm::GridProperty<double>::GridProperty (this=0x7fffffffdbf0, nx=10, ny=10, nz=3, kwInfo=...)
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Grid/GridProperty.cpp:119
#9  0x0000000000e9f0be in Opm::GridProperties<double>::insertKeyword (this=0x19893e0, supportedKeyword=...)
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Grid/GridProperties.cpp:237
#10 0x0000000000e9efaa in Opm::GridProperties<double>::addAutoGeneratedKeyword_ (this=0x19893e0, keywordName="TEMPI")
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Grid/GridProperties.cpp:580
#11 0x0000000000e9ed36 in Opm::GridProperties<double>::getKeyword (this=0x19893e0, keyword="TEMPI")
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Grid/GridProperties.cpp:558
#12 0x0000000000e7772a in Opm::Eclipse3DProperties::getDoubleGridProperty (this=0x1989280, keyword="TEMPI")
    at /home/and/src/opm-parser/lib/eclipse/EclipseState/Eclipse3DProperties.cpp:542
#13 0x0000000000c645cc in Ewoms::EclEquilInitializer<Ewoms::Properties::TTag::EclProblem>::EclEquilInitializer<Opm::EclMaterialLawManager<Opm::ThreePhaseMaterialTraits<double, 0, 1, 2> > > (this=0x7fffffffe030, simulator=..., materialLawManager=...)
    at /home/and/src/ewoms/build-cmake/fake-src/ebos/eclequilinitializer.hh:105
#14 0x0000000000c35272 in Ewoms::EclProblem<Ewoms::Properties::TTag::EclProblem>::readEquilInitialCondition_ (this=0x1928a20)
    at /home/and/src/ewoms/build-cmake/fake-src/ebos/eclproblem.hh:1334
#15 0x0000000000c17f65 in Ewoms::EclProblem<Ewoms::Properties::TTag::EclProblem>::readInitialCondition_ (this=0x1928a20)
    at /home/and/src/ewoms/build-cmake/fake-src/ebos/eclproblem.hh:1324
#16 0x0000000000bfe141 in Ewoms::EclProblem<Ewoms::Properties::TTag::EclProblem>::finishInit (this=0x1928a20)
    at /home/and/src/ewoms/build-cmake/fake-src/ebos/eclproblem.hh:406
#17 0x0000000000bec011 in Ewoms::Simulator<Ewoms::Properties::TTag::EclProblem>::Simulator (this=0x7fffffffe440, verbose=true)
    at /home/and/src/ewoms/build-cmake/fake-src/ewoms/common/simulator.hh:140
#18 0x0000000000b6c975 in Ewoms::start<Ewoms::Properties::TTag::EclProblem> (argc=5, argv=0x7fffffffe778)
    at /home/and/src/ewoms/build-cmake/fake-src/ewoms/common/start.hh:300
#19 0x0000000000b6c329 in main (argc=5, argv=0x7fffffffe778) at /home/and/src/ewoms/build-cmake/fake-src/ebos/ebos.cc:43
```

Reply:
> you are not the most trusting person, are you? ;)

I realize I have been quite anal on this, but the point is that I have been trying quite hard to get people to realize that `xxxNUM` values should be interpreted as ID's and not indices, and you sort of insisted equally hard on considering the value as an index.

Mind you - I am not by any means ruling out the possibility of a bug here; but your suggested fix is wrong. As you can see from the backtrace this blows up because the tablecontainer in questions replies: "Sorry dude - I do not have that many tables" - i.e. the container is *empty*; that is the real bug. I think this is the correct fix:

```diff
diff --git a/lib/eclipse/EclipseState/Tables/TableContainer.cpp b/lib/eclipse/EclipseState/Tables/TableContainer.cpp
index d3f435a..d8f4973 100644
--- a/lib/eclipse/EclipseState/Tables/TableContainer.cpp
+++ b/lib/eclipse/EclipseState/Tables/TableContainer.cpp
@@ -49,7 +49,7 @@ namespace Opm {
 
 
     const SimpleTable& TableContainer::getTable(size_t tableNumber) const {
-        if (tableNumber >= m_maxTables)
+        if (tableNumber > m_maxTables)
             throw std::invalid_argument("TableContainer - invalid tableNumber");
 
         if (hasTable(tableNumber)) {
```

Please also add a small test so we don't trip on this again.

Reply:
don't worry, I can take some rough remarks if you're willing to accept some snippy responses.

looking at `getTable()` it seems like this function mixes up the concepts of IDs and indices itself. That said, I'm very sure that this is not the only occasion where a table is accessed, i.e., if you want to change this to IDs you will have to fix quite a bit of downstream code.

Reply:
> and you sort of insisted equally hard on considering the value as an index.

If I did, I'm sorry, I'm not doing so anymore. just make sure that downsteam does not break.

Reply:
for your fix: what happens if you want to access the table with ID 0?

Reply:
dang - this was more messy than I thought and hoped; might have to fold here :-(

Reply:
> Please also add a small test so we don't trip on this again.

I just wanted to do this, but it is already tested that tables are indexed [0..size()]:

https://github.com/OPM/opm-parser/blob/master/lib/eclipse/tests/TableManagerTests.cpp#L1175-L1176